src/main/java/org/truffleruby/core/module/ModuleFields.java

```diff
@@ -849,8 +857,9 @@ public void getAdjacentObjects(Set<Object> adjacent) {
             ObjectGraph.addProperty(adjacent, constant);
         }
 
-        for (Object value : classVariables.values()) {
-            ObjectGraph.addProperty(adjacent, value);
+        for (Object key : classVariables.getShape().getKeys()) {
+            ObjectGraph
+                    .addProperty(adjacent, DynamicObjectLibrary.getUncached().getOrDefault(classVariables, key, null));
         }
```

Comment:
This should be simplified to `ObjectGraph.addProperty(adjacent, classVariables)`. This will also mark it shared when the RubyModule gets shared, as well as any objects it refers to, which is the correct thing to do.